worker-macros/src/durable_object.rs

```diff
@@ -109,24 +124,123 @@ pub fn expand_macro(tokens: TokenStream) -> syn::Result<TokenStream> {
 
                             #method
                         }
-                    }
+                    },
+                    "on_message" => {
+                        optional_methods.has_on_message = true;
+
+                        let mut method = impl_method.clone();
+                        method.sig.ident = Ident::new("_on_message_raw", method.sig.ident.span());
+                        method.vis = Visibility::Inherited;
+
+                        quote! {
+                            #pound[wasm_bindgen::prelude::wasm_bindgen(js_name = webSocketMessage)]
+                            pub fn _on_message(&mut self, ws: worker_sys::web_sys::WebSocket, message: String) -> js_sys::Promise {
```

Comment:
Binary websocket messages would deliver a message that's an `ArrayBuffer`, so I think this needs to handle both types.

This can be handled by having the incoming type be a `JsValue`, and then that can be converted into an enum depending on its JS type, e.g.
```rust

pub enum WebSocketIncomingMessage {
    String(String),
    Binary(Vec<u8>),
}
```

```rust
let ws_message = if let Some(s) = message.as_string() {
    worker::WebSocketIncomingMessage::String(s)
} else {
    let v = js_sys::Uint8Array::new(&message).to_vec();
    worker::WebSocketIncomingMessage::Binary(v)
};
```
I'm not certain whether the conversions above are infallible, but so far they seem to work for me. (I have a branch where I had been prototyping [a similar change](https://github.com/eric-seppanen/workers-rs/tree/hibernatable).)

Reply:
Good point, I wasn't sure what to make of the `string | ArrayBuffer` type in the docs but I didn't consider binary messages.